Migrate FakeTensors to always call into FakeTensorMode and have them hold a reference

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78972
* #78523
* #78522
* __->__ #78677

Migrates FakeTensors to always invoke FakeTensorMode instead of calling into shared impl function. Because `FakeTensorMode` stores a `FakeTensorConverter`, which holds a map from Real Tensors -> Fake Tensors, this also changes the `tensor_memo` to use a `WeakValueDictionary` to avoid a circular reference from `FakeTensorMode` -> `FakeTensor` -> `FakeTensorMode`... 

When instantiating `FakeTensors` you now need a corresponding `FakeTensorMode` for them to own. If creating new tensors, the idiom is similar to existing fake_mode` from `torchdistx. 
```
with enable_torch_dispatch_mode(FakeTensorMode(inner=None)):
            y = torch.rand([4], device="cpu")
 ```
When converting existing tensors to `FakeTensors`, the idiom is to instantiate a `FakeTensorMode`, and then that to use the conversion, with the new mode associated with the new tensor.

eellison committed:
Migrate FakeTensors to always call into FakeTensorMode and have them hold a reference

[ghstack-poisoned]

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -94,135 +109,50 @@ def __call__(self, t, device=None):
+                if fake_mode is None:
+                    fake_mode = arg.fake_mode
+                else:
+                    assert fake_mode is arg.fake_mode, "Mixing modes NYI"
```
@samdow we should probably do a utility for doing this

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -56,36 +58,49 @@ def _is_tensor_constructor(func: OpOverload):
+
+    def from_real_tensor(self, fake_mode, t):
+        if self._get_memo(t) is not None:
+            return self.tensor_memo[t]
```
This looks wrong shouldn't you return the results of get memo

eellison replied:
wheres the walrus operator where u need it (doesnt work python 3.7)

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -292,6 +222,99 @@ def __init__(self):
+        kwargs = kwargs if kwargs else {}
+
+        # TODO: apply as no_dispatch decorator
+        with no_dispatch():
```
Would be great to have a test suite/ demo / something to test if your new tensor subclass is compositional because it's a bit hard to reason about without actually going through tests/debugging/seeing how things work together. In subclass zoo for example

eellison replied:
Without the no-dispatch I get infinite recursion. the `r = func(*args, **kwargs)` [here](https://github.com/pytorch/pytorch/pull/78677/files#diff-39e82af71afdadbc56a4ecf552ec668ddcda794f8ea3ec41edaef23456fd56e9R295) just calls into `__torch_dispatch__` again

samdow replied:
If we want this to work and it to be compositional with other modes, we'll need to unwrap args and kwargs. Usually this would just be `arg.elem if isinstance(arg, FakeTensor) else arg` but it seems like this will probably be more complicated before it can be a wrapper tensor? Not sure if @ezyang has any thoughts on how to do this

samdow replied:
Ah looks like this should work--so basically we should be saving elem and then unwrapping to that elem, like Ed said, if the mode matches your mode (currently this should always happen since we error if there's FakeTensors from more than one mode but we can be extra careful to set us up for nested FakeTensors if we want)

I got confused by [this comment](https://github.com/pytorch/pytorch/pull/78677/files#diff-39e82af71afdadbc56a4ecf552ec668ddcda794f8ea3ec41edaef23456fd56e9R125) though: `elem does not need to be recorded, because FakeTensor *is a* elem`. FakeTensor should only be taking a vanilla tensor as elem?

samdow replied:
Ah I see--that is Tracer/ProxyTensor specific. And kinda quirky 

Yeah so the idea is that this is basically `FakeTensor(torch.Tensor(...))` and we want to redispatch on `torch.Tensor` instead of on FakeTensor so that we don't get the infinite recursion you're seeing

So if we save `elem`, we should be able to unwrap by basically doing `arg.elem if isinstance(arg, FakeTensor) and self == arg.fake_mode else arg` for every arg and kwarg (this adds in Ed's check that we only unwrap tensors in the same mode) and then run `func(*new_args, **new_kwargs)`

eellison replied:
There is no `elem`, since FakeTensor inherits it's not compositional.. im going to defer this working for another pr

eellison committed:
Update on "Migrate FakeTensors to always call into FakeTensorMode and have them hold a reference"


Migrates FakeTensors to always invoke FakeTensorMode instead of calling into shared impl function. Because `FakeTensorMode` stores a `FakeTensorConverter`, which holds a map from Real Tensors -> Fake Tensors, this also changes the `tensor_memo` to use a `WeakValueDictionary` to avoid a circular reference from `FakeTensorMode` -> `FakeTensor` -> `FakeTensorMode`... 

When instantiating `FakeTensors` you now need a corresponding `FakeTensorMode` for them to own. If creating new tensors, the idiom is similar to existing fake_mode` from `torchdistx. 
```
with enable_torch_dispatch_mode(FakeTensorMode(inner=None)):
            y = torch.rand([4], device="cpu")
 ```
When converting existing tensors to `FakeTensors`, the idiom is to instantiate a `FakeTensorMode`, and then that to use the conversion, with the new mode associated with the new tensor. 

 


[ghstack-poisoned]

eellison committed:
Update on "Migrate FakeTensors to always call into FakeTensorMode and have them hold a reference"


Migrates FakeTensors to always invoke FakeTensorMode instead of calling into shared impl function. Because `FakeTensorMode` stores a `FakeTensorConverter`, which holds a map from Real Tensors -> Fake Tensors, this also changes the `tensor_memo` to use a `WeakValueDictionary` to avoid a circular reference from `FakeTensorMode` -> `FakeTensor` -> `FakeTensorMode`... 

When instantiating `FakeTensors` you now need a corresponding `FakeTensorMode` for them to own. If creating new tensors, the idiom is similar to existing fake_mode` from `torchdistx. 
```
with enable_torch_dispatch_mode(FakeTensorMode(inner=None)):
            y = torch.rand([4], device="cpu")
 ```
When converting existing tensors to `FakeTensors`, the idiom is to instantiate a `FakeTensorMode`, and then that to use the conversion, with the new mode associated with the new tensor. 

 


[ghstack-poisoned]

eellison committed:
Update on "Migrate FakeTensors to always call into FakeTensorMode and have them hold a reference"


Migrates FakeTensors to always invoke FakeTensorMode instead of calling into shared impl function. Because `FakeTensorMode` stores a `FakeTensorConverter`, which holds a map from Real Tensors -> Fake Tensors, this also changes the `tensor_memo` to use a `WeakValueDictionary` to avoid a circular reference from `FakeTensorMode` -> `FakeTensor` -> `FakeTensorMode`... 

When instantiating `FakeTensors` you now need a corresponding `FakeTensorMode` for them to own. If creating new tensors, the idiom is similar to existing fake_mode` from `torchdistx. 
```
with enable_torch_dispatch_mode(FakeTensorMode(inner=None)):
            y = torch.rand([4], device="cpu")
 ```
When converting existing tensors to `FakeTensors`, the idiom is to instantiate a `FakeTensorMode`, and then that to use the conversion, with the new mode associated with the new tensor. 

 


[ghstack-poisoned]

eellison committed:
Update on "Migrate FakeTensors to always call into FakeTensorMode and have them hold a reference"


Migrates FakeTensors to always invoke FakeTensorMode instead of calling into shared impl function. Because `FakeTensorMode` stores a `FakeTensorConverter`, which holds a map from Real Tensors -> Fake Tensors, this also changes the `tensor_memo` to use a `WeakValueDictionary` to avoid a circular reference from `FakeTensorMode` -> `FakeTensor` -> `FakeTensorMode`... 

When instantiating `FakeTensors` you now need a corresponding `FakeTensorMode` for them to own. If creating new tensors, the idiom is similar to existing fake_mode` from `torchdistx. 
```
with enable_torch_dispatch_mode(FakeTensorMode(inner=None)):
            y = torch.rand([4], device="cpu")
 ```
When converting existing tensors to `FakeTensors`, the idiom is to instantiate a `FakeTensorMode`, and then that to use the conversion, with the new mode associated with the new tensor. 

 


[ghstack-poisoned]

eellison committed:
Update on "Migrate FakeTensors to always call into FakeTensorMode and have them hold a reference"


Migrates FakeTensors to always invoke FakeTensorMode instead of calling into shared impl function. Because `FakeTensorMode` stores a `FakeTensorConverter`, which holds a map from Real Tensors -> Fake Tensors, this also changes the `tensor_memo` to use a `WeakValueDictionary` to avoid a circular reference from `FakeTensorMode` -> `FakeTensor` -> `FakeTensorMode`... 

When instantiating `FakeTensors` you now need a corresponding `FakeTensorMode` for them to own. If creating new tensors, the idiom is similar to existing fake_mode` from `torchdistx. 
```
with enable_torch_dispatch_mode(FakeTensorMode(inner=None)):
            y = torch.rand([4], device="cpu")
 ```
When converting existing tensors to `FakeTensors`, the idiom is to instantiate a `FakeTensorMode`, and then that to use the conversion, with the new mode associated with the new tensor. 

 


[ghstack-poisoned]

eellison committed:
Update on "Migrate FakeTensors to always call into FakeTensorMode and have them hold a reference"


Migrates FakeTensors to always invoke FakeTensorMode instead of calling into shared impl function. Because `FakeTensorMode` stores a `FakeTensorConverter`, which holds a map from Real Tensors -> Fake Tensors, this also changes the `tensor_memo` to use a `WeakValueDictionary` to avoid a circular reference from `FakeTensorMode` -> `FakeTensor` -> `FakeTensorMode`... 

When instantiating `FakeTensors` you now need a corresponding `FakeTensorMode` for them to own. If creating new tensors, the idiom is similar to existing fake_mode` from `torchdistx. 
```
with enable_torch_dispatch_mode(FakeTensorMode(inner=None)):
            y = torch.rand([4], device="cpu")
 ```
When converting existing tensors to `FakeTensors`, the idiom is to instantiate a `FakeTensorMode`, and then that to use the conversion, with the new mode associated with the new tensor. 

 


[ghstack-poisoned]

eellison committed:
Update on "Migrate FakeTensors to always call into FakeTensorMode and have them hold a reference"


Migrates FakeTensors to always invoke FakeTensorMode instead of calling into shared impl function. Because `FakeTensorMode` stores a `FakeTensorConverter`, which holds a map from Real Tensors -> Fake Tensors, this also changes the `tensor_memo` to use a `WeakValueDictionary` to avoid a circular reference from `FakeTensorMode` -> `FakeTensor` -> `FakeTensorMode`... 

When instantiating `FakeTensors` you now need a corresponding `FakeTensorMode` for them to own. If creating new tensors, the idiom is similar to existing fake_mode` from `torchdistx. 
```
with enable_torch_dispatch_mode(FakeTensorMode(inner=None)):
            y = torch.rand([4], device="cpu")
 ```
When converting existing tensors to `FakeTensors`, the idiom is to instantiate a `FakeTensorMode`, and then that to use the conversion, with the new mode associated with the new tensor. 

 


[ghstack-poisoned]

eellison committed:
Update on "Migrate FakeTensors to always call into FakeTensorMode and have them hold a reference"


Migrates FakeTensors to always invoke FakeTensorMode instead of calling into shared impl function. Because `FakeTensorMode` stores a `FakeTensorConverter`, which holds a map from Real Tensors -> Fake Tensors, this also changes the `tensor_memo` to use a `WeakValueDictionary` to avoid a circular reference from `FakeTensorMode` -> `FakeTensor` -> `FakeTensorMode`... 

When instantiating `FakeTensors` you now need a corresponding `FakeTensorMode` for them to own. If creating new tensors, the idiom is similar to existing fake_mode` from `torchdistx. 
```
with enable_torch_dispatch_mode(FakeTensorMode(inner=None)):
            y = torch.rand([4], device="cpu")
 ```
When converting existing tensors to `FakeTensors`, the idiom is to instantiate a `FakeTensorMode`, and then that to use the conversion, with the new mode associated with the new tensor. 

 


[ghstack-poisoned]

eellison committed:
Update on "Migrate FakeTensors to always call into FakeTensorMode and have them hold a reference"


Migrates FakeTensors to always invoke FakeTensorMode instead of calling into shared impl function. Because `FakeTensorMode` stores a `FakeTensorConverter`, which holds a map from Real Tensors -> Fake Tensors, this also changes the `tensor_memo` to use a `WeakValueDictionary` to avoid a circular reference from `FakeTensorMode` -> `FakeTensor` -> `FakeTensorMode`... 

When instantiating `FakeTensors` you now need a corresponding `FakeTensorMode` for them to own. If creating new tensors, the idiom is similar to existing fake_mode` from `torchdistx. 
```
with enable_torch_dispatch_mode(FakeTensorMode(inner=None)):
            y = torch.rand([4], device="cpu")
 ```
When converting existing tensors to `FakeTensors`, the idiom is to instantiate a `FakeTensorMode`, and then that to use the conversion, with the new mode associated with the new tensor. 

 


[ghstack-poisoned]

eellison committed:
Update on "Migrate FakeTensors to always call into FakeTensorMode and have them hold a reference"


Migrates FakeTensors to always invoke FakeTensorMode instead of calling into shared impl function. Because `FakeTensorMode` stores a `FakeTensorConverter`, which holds a map from Real Tensors -> Fake Tensors, this also changes the `tensor_memo` to use a `WeakValueDictionary` to avoid a circular reference from `FakeTensorMode` -> `FakeTensor` -> `FakeTensorMode`... 

When instantiating `FakeTensors` you now need a corresponding `FakeTensorMode` for them to own. If creating new tensors, the idiom is similar to existing fake_mode` from `torchdistx. 
```
with enable_torch_dispatch_mode(FakeTensorMode(inner=None)):
            y = torch.rand([4], device="cpu")
 ```
When converting existing tensors to `FakeTensors`, the idiom is to instantiate a `FakeTensorMode`, and then that to use the conversion, with the new mode associated with the new tensor. 

 


[ghstack-poisoned]

eellison committed:
Update on "Migrate FakeTensors to always call into FakeTensorMode and have them hold a reference"


Migrates FakeTensors to always invoke FakeTensorMode instead of calling into shared impl function. Because `FakeTensorMode` stores a `FakeTensorConverter`, which holds a map from Real Tensors -> Fake Tensors, this also changes the `tensor_memo` to use a `WeakValueDictionary` to avoid a circular reference from `FakeTensorMode` -> `FakeTensor` -> `FakeTensorMode`... 

When instantiating `FakeTensors` you now need a corresponding `FakeTensorMode` for them to own. If creating new tensors, the idiom is similar to existing fake_mode` from `torchdistx. 
```
with enable_torch_dispatch_mode(FakeTensorMode(inner=None)):
            y = torch.rand([4], device="cpu")
 ```
When converting existing tensors to `FakeTensors`, the idiom is to instantiate a `FakeTensorMode`, and then that to use the conversion, with the new mode associated with the new tensor. 

 


[ghstack-poisoned]

eellison committed:
Update on "Migrate FakeTensors to always call into FakeTensorMode and have them hold a reference"


Migrates FakeTensors to always invoke FakeTensorMode instead of calling into shared impl function. Because `FakeTensorMode` stores a `FakeTensorConverter`, which holds a map from Real Tensors -> Fake Tensors, this also changes the `tensor_memo` to use a `WeakValueDictionary` to avoid a circular reference from `FakeTensorMode` -> `FakeTensor` -> `FakeTensorMode`... 

When instantiating `FakeTensors` you now need a corresponding `FakeTensorMode` for them to own. If creating new tensors, the idiom is similar to existing fake_mode` from `torchdistx. 
```
with enable_torch_dispatch_mode(FakeTensorMode(inner=None)):
            y = torch.rand([4], device="cpu")
 ```
When converting existing tensors to `FakeTensors`, the idiom is to instantiate a `FakeTensorMode`, and then that to use the conversion, with the new mode associated with the new tensor. 

 


[ghstack-poisoned]

eellison committed:
Update on "Migrate FakeTensors to always call into FakeTensorMode and have them hold a reference"


Migrates FakeTensors to always invoke FakeTensorMode instead of calling into shared impl function. Because `FakeTensorMode` stores a `FakeTensorConverter`, which holds a map from Real Tensors -> Fake Tensors, this also changes the `tensor_memo` to use a `WeakValueDictionary` to avoid a circular reference from `FakeTensorMode` -> `FakeTensor` -> `FakeTensorMode`... 

When instantiating `FakeTensors` you now need a corresponding `FakeTensorMode` for them to own. If creating new tensors, the idiom is similar to existing fake_mode` from `torchdistx. 
```
with enable_torch_dispatch_mode(FakeTensorMode(inner=None)):
            y = torch.rand([4], device="cpu")
 ```
When converting existing tensors to `FakeTensors`, the idiom is to instantiate a `FakeTensorMode`, and then that to use the conversion, with the new mode associated with the new tensor. 

 


[ghstack-poisoned]

eellison commented:
@pytorchbot merge
